server/src/main/java/io/crate/types/StringType.java

```diff
@@ -79,77 +80,87 @@ public class StringType extends DataType<String> implements Streamer<String> {
+                                Object lowerTerm,
+                                Object upperTerm,
+                                boolean includeLower,
+                                boolean includeUpper,
+                                boolean hasDocValues,
+                                boolean isIndexed) {
+            if (isIndexed) {
+                return new TermRangeQuery(
+                    field,
+                    BytesRefs.toBytesRef(preProcess.apply(lowerTerm)),
+                    BytesRefs.toBytesRef(preProcess.apply(upperTerm)),
+                    includeLower,
+                    includeUpper
+                );
             }
+            if (hasDocValues) {
+                return SortedSetDocValuesField.newSlowRangeQuery(
+                    field,
+                    BytesRefs.toBytesRef(preProcess.apply(lowerTerm)),
+                    BytesRefs.toBytesRef(preProcess.apply(upperTerm)),
+                    includeLower,
+                    includeUpper
+                );
+            }
+            return null;
         }
-    ) {
 
         @Override
-        @SuppressWarnings({"rawtypes"})
-        public ValueIndexer<Object> valueIndexer(RelationName table,
-                                                 Reference ref,
-                                                 Function<ColumnIdent, Reference> getRef) {
-            return switch (ref.indexType()) {
-                case FULLTEXT -> (ValueIndexer) new FulltextIndexer(ref);
-                case NONE, PLAIN -> (ValueIndexer) new StringIndexer(ref);
-            };
+        public Query termsQuery(String field, List<Object> nonNullValues, boolean hasDocValues, boolean isIndexed) {
+            if (isIndexed) {
+                return new TermInSetQuery(field, nonNullValues.stream().map(
+                    c -> BytesRefs.toBytesRef(preProcess.apply(c))).toList());
+            }
+            if (hasDocValues) {
+                return SortedSetDocValuesField.newSlowSetQuery(
+                    field, Lists.map(nonNullValues, c -> BytesRefs.toBytesRef(preProcess.apply(c))));
+            }
+            return null;
         }
+    }
+
+    private static final StorageSupport<Object> STORAGE = new StorageSupport<>(
+        true,
+        true,
+        new StringEqQuery(UnaryOperator.identity())) {
+
+            @Override
+            @SuppressWarnings({"rawtypes"})
+            public ValueIndexer<Object> valueIndexer(RelationName table,
+                                                     Reference ref,
+                                                     Function<ColumnIdent, Reference> getRef) {
+                return switch (ref.indexType()) {
+                    case FULLTEXT -> (ValueIndexer) new FulltextIndexer(ref);
+                    case NONE, PLAIN -> (ValueIndexer) new StringIndexer(ref);
+                };
+            }
     };
 
-    private final int lengthLimit;
+    protected int lengthLimit;
```

Comment:
Isn't this still final?

Reply:
hm, not as is, but I can call `super` in the CharacterType ctor, thx!